File: 3-kubernetes/common/mytravels.domain/Features/PointOfInterest/PointOfInterestService.cs
using MetadataExtractor;
using Microsoft.AspNetCore.Http;
using mytravels.contract.CustomException;
using mytravels.contract.Constants;
using mytravels.contract.Dtos;
using mytravels.contract.Interfaces;
using mytravels.contract.Lookups;
using mytravels.contract.Messages;
using mytravels.contract.Responses;
using mytravels.domain.Extensions;

namespace mytravels.domain.Features.PointOfInterest
{
    public class PointOfInterestService : IPointOfInterestService
    {
        private readonly ICoreDbContext _context;
        private readonly IObjectStorageService _objectStorageService;
        private readonly IMessagePublisher _publisher;
        private readonly IGeoService _geoService;

        public PointOfInterestService
        (
            IObjectStorageService service,
            ICoreDbContext context,
            IMessagePublisher publisher,
            IGeoService geoService
        )
        {
            _objectStorageService = service ?? throw new ArgumentNullException(nameof(service));
            _context = context ?? throw new ArgumentNullException(nameof(context));
            _publisher = publisher ?? throw new ArgumentNullException(nameof(publisher));
            _geoService = geoService ?? throw new ArgumentNullException(nameof(geoService));
        }

        public async Task<List<GetPointOfInterestResponse>> GetAsync(CancellationToken cancellationToken)
            => await _context.GetAllPointsOfInterestAsync(cancellationToken);

        public async Task<List<GetPointOfInterestResponse>> GetAsync(string tagName, CancellationToken cancellationToken)
            => await _context.GetPointsOfInterestByTagAsync(tagName, cancellationToken);

        public async Task<int> SaveFileAsPointOfInsterestAsync(IFormFile file, CancellationToken cancellationToken)
        {
            string objectName = await _objectStorageService.SaveObjectAsync(file, BucketNames.NewUploadedImagesContainer, cancellationToken);
            GeoLocation geolocation = await GetCoordinatesAsync(objectName, cancellationToken);

            CreatePointOfInterestDto dto = new()
            {
                OriginalFileName = file.FileName,
                BlobName = objectName,
                FormattedAddress = string.Empty,
                Latitude = geolocation.Latitude,
                Longitude = geolocation.Longitude,
                PointOfInterestTypeId = (int)PointOfInterestTypesEnum.Image
            };

            contract.Entities.PointOfInterest point = dto.ToEntity();
            int id = await _context.CreatePointOfInterestAsync(point, cancellationToken);

            await _publisher.PublishAsync(ExchangeNames.AppendFormattedAddress, new PointOfInterestMessage { CorrelationId = Guid.NewGuid(), PointOfInterestId = id }, cancellationToken);
            await _publisher.PublishAsync(ExchangeNames.ResizeImage, new PointOfInterestMessage { PointOfInterestId = point.Id }, cancellationToken);

            return id;
        }

        public async Task<int> UpdatePointOfInterestAsync(IFormFile file, string pointOfInterestKey, CancellationToken cancellationToken)
        {
            string objectName = await _objectStorageService.SaveObjectAsync(file, BucketNames.NewUploadedImagesContainer, cancellationToken);
            List<contract.Entities.PointOfInterest> points = await _context.GetPointsOfInterestAsync(cancellationToken);
            contract.Entities.PointOfInterest point = points.Where(x => x.PointOfInterestKey == pointOfInterestKey)
                                                   .OrderByDescending(x => x.DateCreated)
                                                   .FirstOrDefault() ?? throw new EntityNotFoundException(nameof(point));

            await _context.AddImageToPointOfInterestAsync(objectName, point, cancellationToken);
            await _publisher.PublishAsync(ExchangeNames.ResizeImage, new PointOfInterestMessage { PointOfInterestId = point.Id }, CancellationToken.None);
            return point.Id;
        }

        public async Task<int> UpdateStatusAsync(string pointOfInterestKey, int pointOfInterestStatusId, CancellationToken cancellationToken)
        {
            List<contract.Entities.PointOfInterest> points = await _context.GetPointsOfInterestAsync(cancellationToken);
            contract.Entities.PointOfInterest point = points.Where(x => x.PointOfInterestKey == pointOfInterestKey)
                                                   .OrderByDescending(x => x.DateCreated)
                                                   .FirstOrDefault() ?? throw new EntityNotFoundException(nameof(point));

            await _context.UpdatePointOfInterestStatusAsync(pointOfInterestStatusId, point, cancellationToken);
            return point.Id;
        }

        private async Task<GeoLocation> GetCoordinatesAsync(string generatedObjectName, CancellationToken cancellationToken)
        {
            Stream savedImage = await _objectStorageService.GetStreamAsync(BucketNames.NewUploadedImagesContainer, generatedObjectName, cancellationToken);
            GeoLocation geolocation = _geoService.ExtractGeoLocation(savedImage);
            if (geolocation.Latitude == 0 || geolocation.Longitude == 0)
                throw new InvalidOperationException("Image is not geocoded");
            return geolocation;
        }
    }
}

File: 1-starter_project_nestjs/libs/domain/src/point-of-interest/point-of-interest.service.ts
import { Injectable, Inject } from '@nestjs/common';
import { InjectDataSource, InjectRepository } from '@nestjs/typeorm';
import { DataSource, Repository } from 'typeorm';
import { v4 as uuidv4 } from 'uuid';
import {
  IPointOfInterestService, IObjectStorageService, IMessagePublisher, IGeoService,
  BucketNames, ExchangeNames, PointOfInterestTypesEnum, PointOfInterestStatusesEnum,
  CreatePointOfInterestDto, GetPointOfInterestResponse, EntityNotFoundException,
  PointOfInterestMessage,
} from '@mytravels/contract';
import { PointOfInterestEntity } from '../entities/point-of-interest.entity';

@Injectable()
export class PointOfInterestService implements IPointOfInterestService {
  constructor(
    @InjectRepository(PointOfInterestEntity)
    private readonly poiRepo: Repository<PointOfInterestEntity>,
    @InjectDataSource()
    private readonly dataSource: DataSource,
    @Inject(IObjectStorageService)
    private readonly storage: IObjectStorageService,
    @Inject(IMessagePublisher)
    private readonly publisher: IMessagePublisher,
    @Inject(IGeoService)
    private readonly geoService: IGeoService,
  ) {}

  async getAll(): Promise<GetPointOfInterestResponse[]> {
    return this.dataSource.query('SELECT * FROM public.spgetpointofinterest()');
  }

  async getByTag(tagName: string): Promise<GetPointOfInterestResponse[]> {
    return this.dataSource.query('SELECT * FROM public.spgetpointofinterestbytagname($1)', [tagName]);
  }

  async saveFileAsPointOfInterest(file: Express.Multer.File): Promise<number> {
    const objectName = await this.storage.saveFile(file, BucketNames.NewUploadedImagesContainer);
    const imageBuffer = await this.storage.getStream(BucketNames.NewUploadedImagesContainer, objectName);
    const geo = await this.geoService.extractGeoLocation(imageBuffer);

    if (!geo.latitude || !geo.longitude) {
      throw new Error('Image is not geocoded');
    }

    const dto: CreatePointOfInterestDto = {
      originalFileName: file.originalname,
      blobName: objectName,
      formattedAddress: '',
      latitude: geo.latitude,
      longitude: geo.longitude,
      pointOfInterestTypeId: PointOfInterestTypesEnum.Image,
    };

    const entity = this.toEntity(dto);
    const saved = await this.poiRepo.save(entity);

    const msg: PointOfInterestMessage = { correlationId: uuidv4(), pointOfInterestId: saved.id };
    await this.publisher.publish(ExchangeNames.AppendFormattedAddress, msg);
    await this.publisher.publish(ExchangeNames.ResizeImage, { correlationId: '', pointOfInterestId: saved.id });

    return saved.id;
  }

  async updatePointOfInterest(file: Express.Multer.File, pointOfInterestKey: string): Promise<number> {
    const objectName = await this.storage.saveFile(file, BucketNames.NewUploadedImagesContainer);
    const existing = await this.poiRepo.findOne({ where: { pointOfInterestKey }, order: { dateCreated: 'DESC' } });
    if (!existing) throw new EntityNotFoundException(pointOfInterestKey);

    const newRow = new PointOfInterestEntity();
    newRow.pointOfInterestKey = existing.pointOfInterestKey;
    newRow.container = BucketNames.NewUploadedImagesContainer;
    newRow.originalFileName = file.originalname;
    newRow.generatedBlobName = objectName;
    newRow.latitude = existing.latitude;
    newRow.longitude = existing.longitude;
    newRow.formattedAddress = existing.formattedAddress;
    newRow.imageResized = false;
    newRow.pointOfInterestTypeId = existing.pointOfInterestTypeId;
    newRow.pointOfInterestStatusId = existing.pointOfInterestStatusId;
    const saved = await this.poiRepo.save(newRow);

    await this.publisher.publish(ExchangeNames.ResizeImage, { correlationId: uuidv4(), pointOfInterestId: saved.id });
    return saved.id;
  }

  async updateStatus(pointOfInterestKey: string, statusId: number): Promise<number> {
    const existing = await this.poiRepo.findOne({ where: { pointOfInterestKey }, order: { dateCreated: 'DESC' } });
    if (!existing) throw new EntityNotFoundException(pointOfInterestKey);

    const newRow = new PointOfInterestEntity();
    newRow.pointOfInterestKey = existing.pointOfInterestKey;
    newRow.container = existing.container;
    newRow.originalFileName = existing.originalFileName;
    newRow.generatedBlobName = existing.generatedBlobName;
    newRow.latitude = existing.latitude;
    newRow.longitude = existing.longitude;
    newRow.formattedAddress = existing.formattedAddress;
    newRow.imageResized = existing.imageResized;
    newRow.pointOfInterestTypeId = existing.pointOfInterestTypeId;
    newRow.pointOfInterestStatusId = statusId;
    newRow.dateUpdated = new Date();
    const saved = await this.poiRepo.save(newRow);
    return saved.id;
  }

  private toEntity(dto: CreatePointOfInterestDto): PointOfInterestEntity {
    const entity = new PointOfInterestEntity();
    entity.originalFileName = dto.originalFileName;
    entity.generatedBlobName = dto.blobName;
    entity.container = BucketNames.NewUploadedImagesContainer;
    entity.dateCreated = new Date();
    entity.latitude = dto.latitude;
    entity.longitude = dto.longitude;
    entity.pointOfInterestTypeId = dto.pointOfInterestTypeId;
    entity.formattedAddress = dto.formattedAddress;
    entity.pointOfInterestKey = uuidv4();
    entity.pointOfInterestStatusId = PointOfInterestStatusesEnum.Open;
    return entity;
  }
}
